backend/src/akvo/lumen/component/keycloak.clj

```diff
@@ -221,11 +239,63 @@
           (println (format "Tried to remove user: %s" user-id))
           (lib/internal-server-error))))))
 
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; API Authorization
+;;;
+
+(defn- active-user [users email]
+  (-> (filter #(and (= (get % "email") email)
+                    (get % "enabled"))
+              users)
+      first
+      (get "id")))
+
+(defn- get-user-id [user-id-cache email]
+  (when-let [user-id (get @user-id-cache email)]
+    (swap! user-id-cache cache/hit email)
+    user-id))
+
+(defn- lookup-user-id
+  "Lookup email -> Keycloak user-id, via cached Keycloak API."
+  [cm req-opts api-root user-id-cache email]
+  (if-some [user-id (get-user-id user-id-cache email)]
+    user-id
+    (let [url (format "%s/users/?email=%s" api-root email)]
+      (let [users (-> (client/get url (assoc req-opts :connection-manager cm)) :body json/decode)]
+        (when-some [user-id (active-user users email)]
+          (-> user-id-cache
+              (swap! assoc email user-id)
+              (get email)))))))
+
+(defn allowed-paths
+  "Provided an email address from the authentication process dig out the
+  Keycloak user and get allowed set of paths, as in #{\"demo/admin\" \"t1\"}
+
+  The Keycloak groups are on the form /akvo/lumen/demo/admin, to simplify  we
+  remove the leading /akvo/lumen and return paths as demo/admin."
+  ([keycloak email]
+   (allowed-paths keycloak email {}))
+  ([{:keys [api-root http-timeout user-id-cache cm] :as keycloak} email
+    {:keys [timeout] :or {timeout http-timeout}}]
+   (let [bare-req-opts {:timeout timeout}]
+     (if-some [headers (request-headers keycloak (assoc bare-req-opts :timeout timeout))]
+       (let [req-opts (assoc bare-req-opts :headers headers)]
+         (when-let [user-id (lookup-user-id cm req-opts api-root user-id-cache email)]
+           (when-let [allowed-paths
```

Comment:
when-let, when-let, if-some, when-let. I wonder if we can simplify this with some macro. I kind of remember somebody publishing some library to avoid this kind of nesting.